### ISSUE TITLE
[Question] How to work with stacks in Docker swarm mode?

### ISSUE THREAD

**Issue by kinghuang** [containous/traefik#804]
I'm deploying complex applications as Docker stacks using `docker stack deploy …`. The current Docker stacks .dab format doesn't provide a way to connect to an external network (docker/docker#25915). There's also no way to add networks to a running swarm service (i.e., add the stack network to a running traefik instance). How do I configure traefik to route to my stack service as the backend in this case?

As it stands, traefik detects my service and configures a backend for it. But, it doesn't have a route to the service because traefik and the stack service aren't on a common overlay network. All I get is a "Bad Gateway" response from traefik.

Is there some way I can configure traefik to use my swarm nodes as the backend servers, and have it pick up the dynamically published service port to connect to the service? This is my first time using traefik, and I'm struggling to figure out how to set this up (or if it's even possible).

For example, here's the info for the main web container in my stack (irrelevant bits excluded).

```json
-bash-4.2$ docker service inspect rms_converis-web
[
    {
        "ID": "daqloiex1e05wimnroc4vz7vs",
        "Version": {
            "Index": 22002
        },
        "CreatedAt": "2016-11-05T02:15:24.614823209Z",
        "UpdatedAt": "2016-11-05T02:53:09.860503578Z",
        "Spec": {
            "Name": "rms_converis-web",
            "Labels": {
                "ca.ucalgary.rms.stage": "dev",
                "com.docker.stack.namespace": "rms",
                "traefik.frontend.entryPoints": "http",
                "traefik.frontend.rule": "Host:master.rms-dev.ucalgary.ca",
                "traefik.port": "80"
            }
        },
        "Endpoint": {
            "Spec": {
                "Mode": "vip",
                "Ports": [
                    {
                        "Protocol": "tcp",
                        "TargetPort": 443
                    },
                    {
                        "Protocol": "tcp",
                        "TargetPort": 80
                    }
                ]
            },
            "Ports": [
                {
                    "Protocol": "tcp",
                    "TargetPort": 443,
                    "PublishedPort": 30000
                },
                {
                    "Protocol": "tcp",
                    "TargetPort": 80,
                    "PublishedPort": 30001
                }
            ],
            "VirtualIPs": [
                {
                    "NetworkID": "68i730xxf68agtuyc2s0e7jz5",
                    "Addr": "10.255.0.11/16"
                }
            ]
        }
    }
]
```

Traefik currently configures a backend with the URL http://10.255.0.11:80. This is not reachable from traefik. What I'd like is for traefik to use my swarm nodes (say 192.168.1.1 and 192.168.1.2) on port 30001 to connect to the service.

**Comment by mvdstam** [comment]
As far as I know, stacks and DAB files are at the time of writing, still experimental. I wouldn't be surprised if a lot of backward-incompatible changes will come in Docker 1.13 regarding stacks. Docker Swarm Mode *kinda* works with one container, and one container only, but I haven't been successful in getting a multi-container application to work (like one would do with docker-compose). It's no surprise that Traefik won't work either here, since the underlying technology is far from finished.

Long story short: unless you're only doing single-container applications, stay away from Docker Swarm-mode for now until it's actually in a stable and ready-to-use state.

**Comment by kinghuang** [comment]
I agree that Docker stacks are considered an experimental feature, subject to change. I am successfully using Docker swarm mode and stacks (generated from Docker Compose files) for my particular system. The only problem I'm having is getting the deployed stacks to work with traefik given the network limitations.

So, my question still stands. Is there a way to label the user-facing web service in such a way that traefik will use the swarm nodes with the ingress networking as backends, instead of the container's internal network and port? I'm guessing the answer is no, as this point?

**Comment by pascalandy** [comment]
Hey @kinghuang,

Can you test with this container? It will display the container ID.

```
ENV_DOMAIN1=iamfoo.example.com
ctn1_NAME=iamfooA
ENV_NET_STATIC=net_static

docker service create \
--name $ctn1_NAME \
--network $ENV_NET_STATIC \
--restart-condition any \
--replicas 1 \
--constraint node.labels.apps_accepted==yes \
--label traefik.frontend.rule=Host:$ENV_DOMAIN1 \
--label traefik.docker.network=$ENV_NET_STATIC \
--label traefik.port=80 \
katacoda/docker-http-server
```

**Comment by kinghuang** [comment]
Hi, @pascalandy. I can run that, though I'm not quite sure what that's supposed to tell me?

I've worked around the limitation with external networks in my particular case by adding steps to the stack deployment that capture the web container's service spec after deployment, remove the service, add a network shared with traefik, and recreating the service.

**Comment by pascalandy** [comment]
> though I'm not quite sure what that's supposed to tell me?
It will display: "This request was processed by host: <ctn-id>"

As I'm reading your case, I think I can't help more than this :-p

Cheers!

**Comment by mgvazquez** [comment]
Same issue here, with 'traefik '1.2 (CodeName: morbier) and Docker 1.13.1

**Comment by pascalandy** [comment]
Please share your docker stack if you expect some help ;-P

**Comment by gvilarino** [comment]
Same issue here

**Comment by sedouard** [comment]
Any way to specify  the `host` name? Problem I'm seeing is that traefik will set the virtual host name as `my-stack-name-myservice.treafikdomain`. I really don't want my stack name associated with the url my app is deployed on :-/

**Comment by sedouard** [comment]
doh - `traefik.frontend.rule` changes that. ignore the above ;-)

**Comment by pascalandy** [comment]
Here is my stack :)
https://github.com/pascalandy/docker-stack-this/tree/master/traefik-manager

As I asked on [Slack](https://traefik.slack.com/archives/C0CDT22PJ/p1497071392758916)

---

Hey Folks!

I just finished a Docker Stack 3.1 compose file `docker swarm mode`. It work perfectly with play-with-docker and on my nodes. It rocks!
 
It runs:
```
1 instance of Traefik
3 instance of caddy
3 instance of caddy
3 instance of who by Emile
```

Now I’m trying to make it works with ACME and I’m stucked. Is there any gents who would enjoy adapting my Docker Stack with ACME and make the TLS magic happen?

Many cheers!
Pascal

https://github.com/pascalandy/docker-stack-this/tree/master/traefik-manager

P.S. This git project also have a setup where Traefik runs on workers nodes.

**Comment by traefiker** [comment]
I'll close the issue due to inactivity but feel free to re-open if needed.

**Comment by pascalandy** [comment]
Here is a solid stack without ACME - https://github.com/pascalandy/docker-stack-this/tree/master/traefik-manager-noacme
